File: scanify/scanify/report/stockist_sales_report/stockist_sales_report.py
import frappe
from frappe import _

def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data

def get_columns():
	return [
		{
			"fieldname": "stockist_code",
			"label": _("Stockist Code"),
			"fieldtype": "Link",
			"options": "Stockist Master",
			"width": 120
		},
		{
			"fieldname": "stockist_name",
			"label": _("Stockist Name"),
			"fieldtype": "Data",
			"width": 200
		},
		{
			"fieldname": "hq",
			"label": _("HQ"),
			"fieldtype": "Link",
			"options": "HQ Master",
			"width": 120
		},
		{
			"fieldname": "team",
			"label": _("Team"),
			"fieldtype": "Link",
			"options": "Team Master",
			"width": 120
		},
		{
			"fieldname": "region",
			"label": _("Region"),
			"fieldtype": "Link",
			"options": "Region Master",
			"width": 120
		},
		{
			"fieldname": "statement_month",
			"label": _("Month"),
			"fieldtype": "Date",
			"width": 100
		},
		{
			"fieldname": "opening_value",
			"label": _("Opening Value"),
			"fieldtype": "Currency",
			"width": 120
		},
		{
			"fieldname": "purchase_value",
			"label": _("Purchase Value"),
			"fieldtype": "Currency",
			"width": 120
		},
		{
			"fieldname": "sales_value",
			"label": _("Sales Value"),
			"fieldtype": "Currency",
			"width": 120
		},
		{
			"fieldname": "closing_value",
			"label": _("Closing Value"),
			"fieldtype": "Currency",
			"width": 120
		}
	]

def get_data(filters):
	conditions = get_conditions(filters)
	
	data = frappe.db.sql(f"""
		SELECT 
			ss.stockist_code,
			ss.stockist_name,
			ss.hq,
			ss.team,
			ss.region,
			ss.statement_month,
			ss.total_opening_value as opening_value,
			ss.total_purchase_value as purchase_value,
			ss.total_sales_value as sales_value,
			ss.total_closing_value as closing_value
		FROM 
			`tabStockist Statement` ss
		WHERE 
			ss.docstatus = 1
			{conditions}
		ORDER BY 
			ss.statement_month DESC, ss.stockist_name
	""", filters, as_dict=1)
	
	return data

def get_conditions(filters):
	conditions = ""
	
	if filters.get("from_date"):
		conditions += " AND ss.statement_month >= %(from_date)s"
	
	if filters.get("to_date"):
		conditions += " AND ss.statement_month <= %(to_date)s"
	
	if filters.get("stockist_code"):
		conditions += " AND ss.stockist_code = %(stockist_code)s"
	
	if filters.get("hq"):
		conditions += " AND ss.hq = %(hq)s"
	
	if filters.get("team"):
		conditions += " AND ss.team = %(team)s"
	
	if filters.get("region"):
		conditions += " AND ss.region = %(region)s"
	
	return conditions

File: scanify/scanify/report/stockist_sales_report/stockist_sales_report.js
frappe.query_reports["Stockist Sales Report"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -6),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "region",
			"label": __("Region"),
			"fieldtype": "Link",
			"options": "Region Master"
		},
		{
			"fieldname": "team",
			"label": __("Team"),
			"fieldtype": "Link",
			"options": "Team Master",
			"get_query": function() {
				let region = frappe.query_report.get_filter_value('region');
				if (region) {
					return {
						"filters": {
							"region": region
						}
					};
				}
			}
		},
		{
			"fieldname": "hq",
			"label": __("HQ"),
			"fieldtype": "Link",
			"options": "HQ Master",
			"get_query": function() {
				let team = frappe.query_report.get_filter_value('team');
				if (team) {
					return {
						"filters": {
							"team": team
						}
					};
				}
			}
		},
		{
			"fieldname": "stockist_code",
			"label": __("Stockist"),
			"fieldtype": "Link",
			"options": "Stockist Master",
			"get_query": function() {
				let hq = frappe.query_report.get_filter_value('hq');
				if (hq) {
					return {
						"filters": {
							"hq": hq
						}
					};
				}
			}
		}
	]
};
